### PR TITLE
CATL-1021: Add Warning Message Instead Of Error When Requirements Not Met

### PR DESCRIPTION
**Pull request by vinuvarshith**
### Overview
https://github.com/compucorp/uk.co.compucorp.civicase/pull/23
In the above PR, we get an error message when 'Drupal' is not used or 'webform_civicrm' doesnt exist. This PR changes that error to a warning message so we dont see an error. 
Also some warnings and notices are handled.

### Before
errormsg

### After
werrormsg

### Technical Details
**Broken Webforms Menu**
In https://github.com/compucorp/uk.co.compucorp.civicase/commit/d88f70bf43152453c4638970bf1ea49f4a884b5d#diff-556c705b947b5cd58c936b775be38e6a the Civicase Case Menu files were reshuffled, but the `isHasSubMenu` function was mistakenly renamed to `hasSubMenu`. 
It has been fixed and `hasSubMenu` is the new name of the function.

Also in `civicase.ang.php`, the Webforms menu has been added to `$options['caseActions']` array instead of `$options`, because `$options['caseActions']` are the options which get displayed for Case Action Menu.

Apart from the fixes above (from @deb1990, in this PR no logic or feature is added or updated. I have fixed/handles warnings and notices. Below is the list of changes made
- api/v3/Case/Getwebforms.php : Error/warning message updated
- CRM/Civicase/Form/CaseWebforms.php : Handle no webforms configured case without notices
- templates/CRM/Civicase/Form/CaseWebforms.tpl : updated to handle error messages better
- ang/civicase.ang.php : fix case when client id is not present
- ang/civicase/case/actions/directives/case-actions.directive.js: handle when no client id is passed from prev step

### CODE REVIEW

**Comment by jamienovick** [comment]
Or we could link to the webform drupal page and state:

! Webform CiviCRM Drupal module is not installed:

- In order to link Drupal Webforms directly from CiviCase you need to install the following Drupal module: [link to Webform_CiviCRM Drupal module].

File: api/v3/Case/Getwebforms.php [hunk 1]
```diff
@@ -1,57 +1,61 @@
...
+ */
+function civicrm_api3_case_getwebforms($params) {
+  $webforms = array();
+  $sysInfo = civicrm_api3('System', 'get')['values'][0];
+  if (!isset($sysInfo['uf']) || $sysInfo['uf'] != 'Drupal') {
+    $out = civicrm_api3_create_success(array());
+    $out['warning_message'] = 'Only Drupal CMS is supported!';
+    return $out;
+  }
+  if (!module_exists('webform_civicrm')) {
```
Comment (deb1990) [inline api/v3/Case/Getwebforms.php]: What did we change in this file? can we make a note of it in the PR descriptions(under technical Details)?

File: api/v3/Case/Getwebforms.php [hunk 2]
```diff
@@ -1,57 +1,61 @@
...
+ * This is used for documentation and validation.
+ *
+ * @param array $spec description of fields supported by this API call
+ *
+ * @return void
+ * 
+ * @see http://wiki.civicrm.org/confluence/display/CRMDOC/API+Architecture+Standards
+ */
+function _civicrm_api3_case_getwebforms_spec(&$spec) {
+  //_civicrm_api3_contact_get_spec($spec);
```
Comment (deb1990) [inline api/v3/Case/Getwebforms.php]: We should not have commented lines, if it is not required please remove it altogether.

**Comment by deb1990** [comment]
@vinuvarshith Please add a technical details section, and add some explanation on the technica approach.

For example you can see the same in https://github.com/compucorp/uk.co.compucorp.civicase/pull/191

File: api/v3/Case/Getwebforms.php (hunk 1 above)
Comment (vinuvarshith) [reply api/v3/Case/Getwebforms.php]: the warning message was updated based on Jamie's comment above

File: api/v3/Case/Getwebforms.php (hunk 2 above)
Comment (vinuvarshith) [reply api/v3/Case/Getwebforms.php]: removed comment

File: ang/civicase.ang.php [hunk 3]
```diff
@@ -203,15 +203,20 @@
       }
       $node = node_load($webform['nid']);
       $data = $node->webform_civicrm['data'];
-      $clients = $data['case'][1]['case'][1]['client_id'];
-      $client = reset($clients);
+      if (isset($data['case'][1]['case'][1]['client_id'])) {
```
Comment (deb1990) [inline ang/civicase.ang.php]: We can simplify this code by
```
$client = 0;
if (isset($data['case'][1]['case'][1]['client_id'])) {
  $clients = $data['case'][1]['case'][1]['client_id'];
  $client = reset($clients);
}

File: ang/civicase/case/actions/directives/case-actions.directive.js [hunk 4]
```diff
@@ -310,11 +310,19 @@
           },
 
           gotoWebform: function (selectedCase, path, clientId) {
-            clientId = 'cid' + clientId;
-            var url = CRM.url(path, {
-              case1: selectedCase.id,
-              [clientId]: selectedCase.client[0].contact_id
-            });
+            var url;
+            if (clientId) {
```
Comment (deb1990) [inline ang/civicase/case/actions/directives/case-actions.directive.js]: there should be an empty line after variable declaration.

File: ang/civicase/case/actions/directives/case-actions.directive.js [hunk 5]
```diff
@@ -310,11 +310,19 @@
...
+              clientId = 'cid' + clientId;
+              url = CRM.url(path, {
+                case1: selectedCase.id,
+                [clientId]: selectedCase.client[0].contact_id
+              });
+            }
+            else {
```
Comment (deb1990) [inline ang/civicase/case/actions/directives/case-actions.directive.js]: There are linting errors in this file, did you run `npm i` in `civicase` folder? If it is installed, you should get errors while commiting files. Please set it up in local, and fix all linting errors.

File: ang/civicase/case/actions/directives/case-actions.directive.js [hunk 6]
```diff
@@ -310,11 +310,19 @@
...
+              url = CRM.url(path, {
+                case1: selectedCase.id
+              });
+            }
             var win = window.open(url, '_blank');
```
Comment (deb1990) [inline ang/civicase/case/actions/directives/case-actions.directive.js]: All variables should be decraled in the first line of the function.

Comment (deb1990) [inline ang/civicase/case/actions/directives/case-actions.directive.js]: `window` object should not be directly used in code. Instead `$window` needs to be injected, and used.

File: ang/civicase/case/actions/directives/case-actions.directive.js [hunk 7]
```diff
@@ -310,11 +310,19 @@
...
-            var url = CRM.url(path, {
-              case1: selectedCase.id,
-              [clientId]: selectedCase.client[0].contact_id
-            });
+            var url;
+            if (clientId) {
+              clientId = 'cid' + clientId;
+              url = CRM.url(path, {
+                case1: selectedCase.id,
+                [clientId]: selectedCase.client[0].contact_id
```
Comment (deb1990) [inline ang/civicase/case/actions/directives/case-actions.directive.js]: `[clientId]` is an ES6 command, and we are not using ES6 yet. So please use ES5 statements.

File: ang/civicase/case/actions/directives/case-actions.directive.js [hunk 8]
```diff
@@ -310,11 +310,19 @@
           },
 
           gotoWebform: function (selectedCase, path, clientId) {
-            clientId = 'cid' + clientId;
-            var url = CRM.url(path, {
-              case1: selectedCase.id,
-              [clientId]: selectedCase.client[0].contact_id
-            });
+            var url;
```
Comment (deb1990) [inline ang/civicase/case/actions/directives/case-actions.directive.js]: Also this function lacks documentation. And unit test. Even if we dont want to add unit test now, documentation must be added.

File: ang/civicase/case/actions/directives/case-actions.directive.js [hunk 9]
```diff
@@ -310,11 +310,19 @@
...
+              url = CRM.url(path, {
+                case1: selectedCase.id,
+                [clientId]: selectedCase.client[0].contact_id
+              });
+            }
+            else {
+              url = CRM.url(path, {
+                case1: selectedCase.id
+              });
+            }
```
Comment (deb1990) [inline ang/civicase/case/actions/directives/case-actions.directive.js]: This is how the whole function should look like

```
/**
 * Opens the Webform in a new tab
 *
 * @param {Object} selectedCase
 * @param {String} path
 * @param {Int/String} clientId
 */
 gotoWebform: function (selectedCase, path, clientId) {
   var window;
   var urlObject = { case1: selectedCase.id };

   if (clientId) {
     urlObject['cid' + clientId] = selectedCase.client[0].contact_id;
   }

   window = $window.open(CRM.url(path, urlObject), '_blank');
   window.focus();
 }
```

**Pull request merged**